pgxn/neon/neon.c

```diff
@@ -522,29 +539,44 @@ RestoreRunningXactsFromClog(CheckPoint *checkpoint, TransactionId **xids, int *n
 			elog(LOG, "too many running xacts to restore from the CLOG; oldestXid=%u oldestActiveXid=%u nextXid %u",
 				 checkpoint->oldestXid, checkpoint->oldestActiveXid,
 				 XidFromFullTransactionId(checkpoint->nextXid));
-			goto fail;
+
+			overflow = true;
+			switch (running_xacts_overflow_policy)
+			{
+				case OP_WAIT:
+					goto fail;
+				case OP_IGNORE:
+					break;
+				case OP_SKIP:
+					n_restored_xids = 0;
+					break;
+			}
 		}
 
 		restored_xids[n_restored_xids++] = xid;
 
 	skip:
 		TransactionIdAdvance(xid);
-		continue;
 	}
 
 	/* sanity check */
 	if (next_prepared_idx != n_prepared_xids)
 	{
 		elog(LOG, "prepared transaction ID %u was not visited in the CLOG scan, cannot restore running-xacts from CLOG",
 			 prepared_xids[next_prepared_idx]);
-		Assert(false);
-		goto fail;
+		if (!overflow)
+		{
+			Assert(false);
+			goto fail;
+		}
 	}
 
 	elog(LOG, "restored %d running xacts by scanning the CLOG; oldestXid=%u oldestActiveXid=%u nextXid %u",
 		 n_restored_xids, checkpoint->oldestXid, checkpoint->oldestActiveXid, XidFromFullTransactionId(checkpoint->nextXid));
 	*nxids = n_restored_xids;
 	*xids = restored_xids;
+	if (prepared_xids)
+		pfree(prepared_xids);
```

Comment:
Can you elaborate on this change? Was it missing in the previous PR? Not sure I understand how it's related to the overflow policy

Reply:
Yes, it was missed in previous PR. It is very minor memory leak (because memory context in any case released.
I have added it for consistency: if it is removed in case of error, then we should also do it in normal case.